doc/lxc.container.conf.sgml.in

```diff
@@ -1503,6 +1503,71 @@ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
           </listitem>
         </varlistentry>
 
+        <varlistentry>
+          <term>
+            <option>lxc.rootfs.overlay.upperdir</option>
+          </term>
+          <listitem>
+            <para>
+              this option can be used as an alternative to lxc.rootfs.path for specifying
+              the upperdir for the rootfs overlay mount. It must be used in conjunction
+              with lxc.rootfs.overlay.lower.[i].path.
+            </para>
+              a pathname is used for this option, for example
+              <filename>/home/joe/.local/share/lxc/my-container/upper</filename>
+            <para>
+            </para>
+          </listitem>
+        </varlistentry>
+
+        <varlistentry>
+          <term>
+            <option>lxc.rootfs.overlay.lower.[i].path</option>
+          </term>
+          <listitem>
+            <para>
+              specify the lower path for the ith lowerdir in the overlay
+            </para>
+            <para>
+              For specifying a directory, a pathname can be used. For a fuse
+              mount, the following syntax is used:
+              <filename>fuse:fuse_binary:[arg_1]:[arg_2]:...:[arg_n]:source</filename>
```

Comment:
I'm curious as to why you chose ':' to separate tokens. 
I'm not really versed in lxc config language.  Is there a reason you chose ":" to delimit tokens?

Reply:
Because ':' is what's been used for probably a decade in lxc.

We did however consider splitting this up, so that it would be

lxc.rootfs.under.0.type = fuse:/usr/bin/squashfuse ( || loop || dir)
lxc.rootfs.under.0.path = filename
lxc.rootfs.under.0.options = allow-other

Reply:
No good reason, I could probably take the entire command line as-is, with regular spaces as the delimiters.

Reply:
@hallyn one thing to consider if we get rid of the colon-separated options is how we pass the `mount` argument to `puzzlefs`. `squashfuse` doesn't need it.

Reply:
Ah, yes.  Same with atomfs.

Reply:
Well, so maybe we just give an "out", like

```
lxc.rootfs.under.0.cmd = /usr/bin/puzzlefs mount /var/lib/lxc/c1/oci imagename $dest
```

Note that '$dest' would be filled in by lxc to the tempdir which it is using for that underlay.

This still doesn't quite help with dmverity.  There we need to actually call cryptsetup in advance and pass the result to the actual mount command to use.  Maybe that's where we draw the line and say "use a pre-mount hook".

Reply:
A comment on this, though.
If I were going to write a 'lxc-oci' template that needed to be able to support both
 * verity mount as root
 * non-verity and fuse (squashfuse) mount as non-root

I would probably find it easier to write a program and use it in both cases via 'lxc.hook.pre-mount' and not use any of the newly added function.

The new 'overlay' support wouldnt really help me.

Reply:
> 
> If the mount program does it, then it's conceivable that you create a root owned container which could do verity mounts and fail on corruption, but the attacker changes environment just enough to make the mount program think it's just running unprivileged and should ignore verity failures.
> 
> So instead the template should say "if i'm root, then create a config that will force use of verity and fail when that fails. If I'm not root, use fuse and accept verity can't work".

Thats a good point. So say there are 2 programs:
 * lxc-oci (the template)
 * lxc-oci-mount - the mounter program

I think I would make 'lxc-oci' write into the generated lxc config  a parameter indicating whether or not it should use verity.

```
lxc.hook.pre-mount = lxc-oci-mount --verity=true oci:/path/to/ocidir:name
```

and then we could add a parameter to the oci template to cause it to write '--verity=true' or '--verity=false'.  the default would be "if root, use verity".

that would suffice, right?